From 2344e747a7dd5e412e38eec23258e848c30a281c Mon Sep 17 00:00:00 2001
From: Michal Kubecek <mkubecek@suse.cz>
Date: Tue, 10 Mar 2020 14:16:01 +0100
Subject: [PATCH 06/11] netlink: add netlink handler for gprivflags
 (--show-priv-flags)

Implement "ethtool --show-priv-flags <dev>" subcommand using
ETHTOOL_MSG_PRIVFLAGS_GET netlink message. This retrieves and displays
values of device private flags, traditionally provided by ETHTOOL_GPFLAGS
ioctl request.

Also register the callback with monitor code so that the monitor can
display ETHTOOL_MSG_PRIVFLAGS_NTF notifications.

Signed-off-by: Michal Kubecek <mkubecek@suse.cz>
---
 Makefile.am         |   2 +-
 ethtool.c           |   1 +
 netlink/extapi.h    |   2 +
 netlink/monitor.c   |   8 ++++
 netlink/netlink.h   |   1 +
 netlink/privflags.c | 101 ++++++++++++++++++++++++++++++++++++++++++++
 6 files changed, 114 insertions(+), 1 deletion(-)
 create mode 100644 netlink/privflags.c

diff --git a/Makefile.am b/Makefile.am
index a4446a8e5681..e55d781c3bf3 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -31,7 +31,7 @@ ethtool_SOURCES += \
 		  netlink/monitor.c netlink/bitset.c netlink/bitset.h \
 		  netlink/settings.c netlink/parser.c netlink/parser.h \
 		  netlink/permaddr.c netlink/prettymsg.c netlink/prettymsg.h \
-		  netlink/features.c \
+		  netlink/features.c netlink/privflags.c \
 		  netlink/desc-ethtool.c netlink/desc-genlctrl.c \
 		  netlink/desc-rtnl.c \
 		  uapi/linux/ethtool_netlink.h \
diff --git a/ethtool.c b/ethtool.c
index f26a68bdb894..d4171ab3abec 100644
--- a/ethtool.c
+++ b/ethtool.c
@@ -5361,6 +5361,7 @@ static const struct option args[] = {
 	{
 		.opts	= "--show-priv-flags",
 		.func	= do_gprivflags,
+		.nlfunc	= nl_gprivflags,
 		.help	= "Query private flags"
 	},
 	{
diff --git a/netlink/extapi.h b/netlink/extapi.h
index bc535ac60a73..e2caa0e269d9 100644
--- a/netlink/extapi.h
+++ b/netlink/extapi.h
@@ -20,6 +20,7 @@ int nl_sset(struct cmd_context *ctx);
 int nl_permaddr(struct cmd_context *ctx);
 int nl_gfeatures(struct cmd_context *ctx);
 int nl_sfeatures(struct cmd_context *ctx);
+int nl_gprivflags(struct cmd_context *ctx);
 int nl_monitor(struct cmd_context *ctx);
 
 void nl_monitor_usage(void);
@@ -44,6 +45,7 @@ static inline void nl_monitor_usage(void)
 #define nl_permaddr		NULL
 #define nl_gfeatures		NULL
 #define nl_sfeatures		NULL
+#define nl_gprivflags		NULL
 
 #endif /* ETHTOOL_ENABLE_NETLINK */
 
diff --git a/netlink/monitor.c b/netlink/monitor.c
index 79f36156af16..0eda27ffe22e 100644
--- a/netlink/monitor.c
+++ b/netlink/monitor.c
@@ -35,6 +35,10 @@ static struct {
 		.cmd	= ETHTOOL_MSG_FEATURES_NTF,
 		.cb	= features_reply_cb,
 	},
+	{
+		.cmd	= ETHTOOL_MSG_PRIVFLAGS_NTF,
+		.cb	= privflags_reply_cb,
+	},
 };
 
 static void clear_filter(struct nl_context *nlctx)
@@ -110,6 +114,10 @@ static struct monitor_option monitor_opts[] = {
 		.pattern	= "-k|--show-features|--show-offload|-K|--features|--offload",
 		.cmd		= ETHTOOL_MSG_FEATURES_NTF,
 	},
+	{
+		.pattern	= "--show-priv-flags|--set-priv-flags",
+		.cmd		= ETHTOOL_MSG_PRIVFLAGS_NTF,
+	},
 };
 
 static bool pattern_match(const char *s, const char *pattern)
diff --git a/netlink/netlink.h b/netlink/netlink.h
index ec9b0be9a161..98c21a967ea1 100644
--- a/netlink/netlink.h
+++ b/netlink/netlink.h
@@ -57,6 +57,7 @@ int linkinfo_reply_cb(const struct nlmsghdr *nlhdr, void *data);
 int wol_reply_cb(const struct nlmsghdr *nlhdr, void *data);
 int debug_reply_cb(const struct nlmsghdr *nlhdr, void *data);
 int features_reply_cb(const struct nlmsghdr *nlhdr, void *data);
+int privflags_reply_cb(const struct nlmsghdr *nlhdr, void *data);
 
 static inline void copy_devname(char *dst, const char *src)
 {
diff --git a/netlink/privflags.c b/netlink/privflags.c
new file mode 100644
index 000000000000..14aca0f8cfc5
--- /dev/null
+++ b/netlink/privflags.c
@@ -0,0 +1,101 @@
+/*
+ * privflags.c - netlink implementation of private flags commands
+ *
+ * Implementation of "ethtool --show-priv-flags <dev>"
+ */
+
+#include <errno.h>
+#include <string.h>
+#include <stdio.h>
+
+#include "../internal.h"
+#include "../common.h"
+#include "netlink.h"
+#include "strset.h"
+#include "bitset.h"
+
+/* PRIVFLAGS_GET */
+
+static void privflags_maxlen_walk_cb(unsigned int idx, const char *name,
+				     bool val, void *data)
+{
+	unsigned int *maxlen = data;
+	unsigned int len, n;
+
+	if (name)
+		len = strlen(name);
+	else {
+		len = 3; /* strlen("bit") */
+		for (n = idx ?: 1; n; n /= 10)
+			len++; /* plus number of ditigs */
+	}
+	if (len > *maxlen)
+		*maxlen = len;
+}
+
+static void privflags_dump_walk_cb(unsigned int idx, const char *name, bool val,
+				   void *data)
+{
+	unsigned int *maxlen = data;
+	char buff[16];
+
+	if (!name) {
+		snprintf(buff, sizeof(buff) - 1, "bit%u", idx);
+		name = buff;
+	}
+	printf("%-*s: %s\n", *maxlen, name, val ? "on" : "off");
+}
+
+int privflags_reply_cb(const struct nlmsghdr *nlhdr, void *data)
+{
+	const struct nlattr *tb[ETHTOOL_A_PRIVFLAGS_MAX + 1] = {};
+	DECLARE_ATTR_TB_INFO(tb);
+	const struct stringset *flag_names = NULL;
+	struct nl_context *nlctx = data;
+	unsigned int maxlen = 0;
+	bool silent;
+	int err_ret;
+	int ret;
+
+	silent = nlctx->is_dump || nlctx->is_monitor;
+	err_ret = silent ? MNL_CB_OK : MNL_CB_ERROR;
+
+	ret = mnl_attr_parse(nlhdr, GENL_HDRLEN, attr_cb, &tb_info);
+	if (ret < 0 || !tb[ETHTOOL_A_PRIVFLAGS_FLAGS])
+		return err_ret;
+	nlctx->devname = get_dev_name(tb[ETHTOOL_A_PRIVFLAGS_HEADER]);
+	if (!dev_ok(nlctx))
+		return MNL_CB_OK;
+
+	if (bitset_is_compact(tb[ETHTOOL_A_PRIVFLAGS_FLAGS])) {
+		ret = netlink_init_ethnl2_socket(nlctx);
+		if (ret < 0)
+			return err_ret;
+		flag_names = perdev_stringset(nlctx->devname, ETH_SS_PRIV_FLAGS,
+					      nlctx->ethnl2_socket);
+	}
+
+	ret = walk_bitset(tb[ETHTOOL_A_PRIVFLAGS_FLAGS], flag_names,
+			  privflags_maxlen_walk_cb, &maxlen);
+	if (ret < 0)
+		return err_ret;
+	if (silent)
+		putchar('\n');
+	printf("Private flags for %s:\n", nlctx->devname);
+	ret = walk_bitset(tb[ETHTOOL_A_PRIVFLAGS_FLAGS], flag_names,
+			  privflags_dump_walk_cb, &maxlen);
+	return (ret < 0) ? err_ret : MNL_CB_OK;
+}
+
+int nl_gprivflags(struct cmd_context *ctx)
+{
+	struct nl_context *nlctx = ctx->nlctx;
+	struct nl_socket *nlsk = nlctx->ethnl_socket;
+	int ret;
+
+	ret = nlsock_prep_get_request(nlsk, ETHTOOL_MSG_PRIVFLAGS_GET,
+				      ETHTOOL_A_PRIVFLAGS_HEADER, 0);
+	if (ret < 0)
+		return ret;
+	return nlsock_send_get_request(nlsk, privflags_reply_cb);
+}
-- 
2.25.1

